OSDOCS-10930:ANP best practices (#77593)
openshift/openshift-docs: modules/nw-ovn-k-anp-best-practices.adoc

/ Module included in the following assemblies:
//
// * list of assemblies where this module is included:
// networking/network_security/AdminNetworkPolicy/nw-ovn-k-anp-recommended-practices.adoc

:_mod-docs-content-type: REFERENCE
[id="anp-best-practices_{context}"]
= Designing AdminNetworkPolicy

When building `AdminNetworkPolicy` (ANP), you should consider the following when creating your policies:

//This will need to be added in at 4.16.1

//* You should create `AdminNetworkPolicy` at different priorities so that precedence is well defined. It is not guaranteed which policy will take precedence when overlapping ANP are created.

> * There is only one `BaselineAdminNetworkPolicy` for the cluster so be careful in defining it and you might consider using it as a default guardrail policy.

Reviewer: I think "there is only one" and "be careful" are separate pieces of advice here:

1. There is only one BANP, and especially with the current version of the API which doesn't have "tenancy", there are only a handful useful things you can do with it, and it would probably help if we explained them:
   - You can set a "default deny" policy for ingress in user namespaces, so that users have to add NetworkPolicies to allow the ingress traffic that they want to allow, and if they don't do that it will be denied.
   - You can set a "default deny" policy for egress in user namespaces, so that users have to add NetworkPolicies to allow the egress traffic that they want to allow, and if they don't do that it will be denied.
     - The above, but also with an "allow" rule to allow egress to DNS by default so users don't have to remember to do that themselves.
   - You can set an egress policy that allows egress to all pods and denies access to `0.0.0.0/0` and `::/0` in user namespaces, so that users still have allow-egress-by-default for cluster-internal traffic, but have deny-by-default for cluster egress traffic, and they have to add NetworkPolicies to allow the cluster egress traffic that they want to allow.
If you don't want to do one of those things, then you _probably_ don't want to use BANP.

2. "Be careful" that you scope your BANP so that it only denies traffic in _user_ namespaces, not _system_ namespaces, because the system namespaces aren't going to have NPs to override your BANP. (Um... I'm not sure exactly how you do this? You need to select on the existence of some label which is part of the default project template...)

Author: Ok, I tried to work this out in some rewording of a new section. Dan, thank you for review.

* Since `0-100` is the supported priority range, it is always better when designing your ANP to use a middle range like `30-70` because this leaves some placeholder
for priorities before and after. Even in the middle range, you might want to leave gaps so that as your infrastructure requirements evolve over time, you are able to insert new ANPs when needed at the right priority level. If you pack your ANPs, then you might need to recreate all of them to accommodate any changes in the future.

> * `networks` peer in egress rule is meant to select external destinations. It is recommended to use `namespaces` and `pods` peers to select in-cluster peers. Do not use `networks` for podIP level matching.

Reviewer: I don't think this is a useful/correct recommendation. The reason `networks` is not recommended for pod IP matching is because it's not _useful_ for that, and this should be completely obvious without us needing to point it out. But if the admin has some strange use case where for some reason using `networks` works better than `namespaces` or `pods`, then they should do that, and it is supported.

Author: So just remove this?

> * When using `0.0.0.0/0` to create a strong `DENY` policy, it is recommended that you check to make sure that you have higher priority `Allow` rules for essential traffic like towards DNS and KAPI-server. This ensures that you do not lock up your entire cluster leaving it in a non-functional state.

Reviewer: "`0.0.0.0/0` or `::/0`"

"`Deny`" should be initial-caps only because that's how it is in the API.

> * When using `0.0.0.0/0` to create a strong `DENY` policy, it is recommended that you check to make sure that you have higher priority `Allow` rules for essential traffic like towards DNS and KAPI-server. This ensures that you do not lock up your entire cluster leaving it in a non-functional state.

Reviewer: re "lock up your entire cluster leaving it in a non-functional state", again, we need to be recommending that admins only create rules that apply to _user_ namespaces, because we do not support them making rules that apply to system namespaces, and it's that case where we need to warn about leaving the cluster in a non-functional state. If they are only applying rules to user namespaces then it doesn't matter if they make them overly restrictive; that will only break user workloads, not the cluster as a whole.

Author: I've tried to clarify this based on my understanding of the feature.

> * `Pass` is a special action used to delegate policy decision to the `NetworkPolicy` layer. If you might consider using this action when considering the interoperability between ANP and your `NetworkPolicy`.

Reviewer: This isn't a "best practice", it's just part of the API definition...
(Also, the second sentence isn't grammatical.)

Author: I _think_ the `You might consider` makes a suggestion that moves in the direction of a best practice, but point taken.

> * Each rule can have up to 100 peers. Every peer is added to the same `Address_Set` object. Selectors across multiple rules should not overlap. This ensures that that the same IPs do not appear in multiple `Address_Set`, which can cause performance and scale scale limitations due to an OVN-Kubernetes limitation.

Reviewer: This is mixing user-level recommendations with internal implementation details. We should not be talking about Address_Sets here.

Author: Try to reword.

> * Using `namedPorts` is not performance efficient since each name has to get evaluated against the `subject` of the policy and `peer` pod's container ports. You might use this with caution in larger clusters where you have many pods and each pod has many containers.

Reviewer: Again, this is mixing in too much low-level information. People don't need to know the low-level details of why it is inefficient. Just tell them what sorts of policies it is OK to use namedPorts in, and what kinds of policies it is not OK, and just _how_ not OK it is. (Is it 10 times less efficient? 10,000 times?)

"has many containers" does not seem relevant at all.

Author: @tssurya this one I definitely need help with. cc @asood-rh

Reviewer 2: Using ports (namedport) creates 3 ACLs, one per per protocol, 3ACLs one per protocol if ports (PortNumber, PortRange) are specified in the ANP. If both are used, results in 6 ACLs. So use wisely.

Author: @danwinship I'm not sure I know how we could even figure out how less efficient it is. I suppose the point here is more along the lines of what @asood-rh is saying which is it isn't a good idea to use `namedPorts` in conjunction with `PortNumber` and `PortRange`. I'm going to update this along those lines for another round of feedback. It may be that we should just take this out.

[id="differences-with-netpol_{context}"]
= Differences to consider between NetworkPolicy and ANP

> * API semantics for ANP are very different from `NetworkPolicy`. There is no implicit deny for ANP when you create the policy. You have to explicitly define what you need. As a consequence, you cannot use ANP to isolate a pod from everything for ingress traffic. You You might consider using `Deny` action to `0.0.0.0/0` and it will act like a default deny to everything or you could `Deny` from all other pods for ingress traffic.

Reviewer: It's not going to be obvious to people how "no default deny" relates to "can't be used to isolate from everything". The obvious response is "but why can't I just make a rule to explicitly deny everything?" This needs to be split up into two (or more) points.

> * API semantics for ANP are very different from `NetworkPolicy`. There is no implicit deny for ANP when you create the policy. You have to explicitly define what you need. As a consequence, you cannot use ANP to isolate a pod from everything for ingress traffic. You You might consider using `Deny` action to `0.0.0.0/0` and it will act like a default deny to everything or you could `Deny` from all other pods for ingress traffic.

Reviewer: Pretty much anywhere we way say "you could make a policy to do X", we should provide an example policy. This is new stuff and there are not a lot of good examples on the web, and half of the examples that people might find are using outdated versions of the API.

Author: I can xref to a policy in the docs that uses a `Deny` action to `0.0.0.0/0`. Does that work?

> * Use the empty namespace selector with care for ANP because it will select _ALL_ namespaces in your cluster including the infrastructure namespaces like `kube-system`, `kapi-server`, and `openshift-*`. To secure your workloads, you might use workload specific labels and then use that as the namespace selector instead of using `{}` string. An empty namespace selector has the potential to lock up your entire cluster leaving it in a non-functional state.

Reviewer: This is not even "with care". We _do not support_ applying ANPs to system namespaces, period. You are virtually gauranteed to eventually break you cluster if you do so.

Author: I tried to address this with a callout and rewording.

> / Module included in the following assemblies:

Reviewer 3: nit: this is showing up in the preview, not sure if it was intended to

> / Module included in the following assemblies:

Reviewer 3: nit: this is showing up in the preview, not sure if it was intended to

Author: thank you.

//
// * list of assemblies where this module is included:
// networking/network_security/AdminNetworkPolicy/nw-ovn-k-anp-recommended-practices.adoc

:_mod-docs-content-type: REFERENCE
[id="anp-best-practices_{context}"]
= Designing AdminNetworkPolicy

When building `AdminNetworkPolicy` (ANP), you should consider the following when creating your policies:

//This will need to be added in at 4.16.1

//* You should create `AdminNetworkPolicy` at different priorities so that precedence is well defined. It is not guaranteed which policy will take precedence when overlapping ANP are created.
* Administrators should build ANP that apply to user namespaces not system namespaces.

[IMPORTANT]
====

> Applying ANP and BANP to system namespaces is not supported because this can leave your cluster unresponsive and in a non-functional state.

Reviewer: ```suggestion
Applying ANP and BANP to system namespaces (`default`, `kube-system`, any namespace whose name starts with `openshift-`, etc) is not supported, and this can leave your cluster unresponsive and in a non-functional state.
```

(changing "because" to "and", because there are other reasons this is not supported too)

====

* Since `0-100` is the supported priority range, it is always better when designing your ANP to use a middle range like `30-70` because this leaves some placeholder
for priorities before and after. Even in the middle range, you might want to leave gaps so that as your infrastructure requirements evolve over time, you are able to insert new ANPs when needed at the right priority level. If you pack your ANPs, then you might need to recreate all of them to accommodate any changes in the future.

> * When using `0.0.0.0/0` or `::/0` to create a strong `Deny` policy, it is recommended that you check to make sure that you have higher priority `Allow` rules for essential traffic.

Reviewer: ```suggestion
* When using `0.0.0.0/0` or `::/0` to create a strong `Deny` policy, it is recommended that you check to make sure that you have higher priority `Allow` or `Pass` rules for essential traffic.
```

> * `Pass` is a special action used to delegate policy decision to the `NetworkPolicy` layer. You might consider using this action when considering the interoperability between ANP and your `NetworkPolicy`.

Reviewer: IMO that's wrong and we should be saying something like.

    Note that an `Allow` rule in an ANP means that the connection will always be allowed, and
    `NetworkPolicy` will be ignored. Use `Allow` when you want to ensure that a connection is
    allowed no matter what. Use `Pass` if you just want to "not deny" the connection, and
    allow the users to decide for themselves whether to allow or deny the connection according
    to their own `NetworkPolicy` objects.

Author: I've tried to clarify this a little by separating it into two points. I understand that the connection you are trying to make is that these actions (pass and allow) are related. However, the actions we are asking the user to take are buried in the current sentence structure. If this doesn't work I can combine.

> * Selectors across multiple rules should not overlap. This ensures that that the same IPs do not appear in multiple policies, which can cause performance and scale limitations due to an OVN-Kubernetes limitation.

Reviewer 3: ```suggestion
* Selectors across multiple rules should not overlap. This ensures that that the same IPs do not appear in multiple policies, which can cause performance and scale limitations.
```

I wouldn't reference and ovn-k limitation here :)

* Using `namedPorts` is not performance efficient since each name has to get evaluated against the `subject` of the policy and `peer` pod's container ports. You might use this with caution in larger clusters where you have many pods and each pod has many containers.

[id="considerations-using-banp_{context}"]
== Considerations for using BaselineAdminNetworkPolicy

> * There is only one `BaselineAdminNetworkPolicy` (BANP) for the cluster be careful in defining it. The following are supported uses for BANP that administrators might consider in designing their BANP:

Reviewer 3: ```suggestion
* Only a single `BaselineAdminNetworkPolicy` (BANP) may be defined within a cluster, make sure the architect it thoughtfully. The following are supported uses for BANP that administrators might consider in designing their BANP:
```

> ** You can set a default deny policy for ingress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the ingress traffic that they want to allow, and if they do not add network policies for ingress it will be denied.

Reviewer 3: ```suggestion
** You can set a default deny policy for cluster-local ingress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the ingress traffic that they want to allow, and if they do not add network policies for ingress it will be denied.
```

Or Something to signify that this applies only to "in-cluster" traffic

> ** You can set a default deny policy for egress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the egress traffic that they want to allow, and if they do not add network policies it will be denied.

Reviewer 3: ```suggestion
** You can set a default deny policy for cluster-local egress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the egress traffic that they want to allow, and if they do not add network policies it will be denied.
```

> ** You can set a default allow policy for egress to DNS.This BANP will make it so that users in namespaces do not have to set an allow egress `NetworkPolicy` to DNS.

Reviewer 3: ```suggestion
** You can set a default allow policy for egress to the in-cluster DNS service. Such a BANP will ensure that namespaced users do not have to set an allow egress `NetworkPolicy` to the in-cluster DNS service.
```

> ** You can set a default allow policy for egress to DNS.This BANP will make it so that users in namespaces do not have to set an allow egress `NetworkPolicy` to DNS.

Reviewer: (This is only interesting in combination with the previous "default-deny egress" policy; if you don't have a default-deny egress policy, you don't need an allow-egress-to-DNS policy because it will be allowed anyway.)

> ** You can set an egress policy that allows egress to all pods and denies access to `0.0.0.0/0` and `::/0` in user namespaces. This BANP will make it so users still have to allow egress by default for any internal traffic in the cluster, but also have to deny by default any egress traffic in the cluster, and they have to add `NetworkPolicy` to allow the cluster egress traffic that they want to allow.

Reviewer 3: ```suggestion
** You can set an egress policy that allows internal egress traffic to all pods but denies access to all external endpoints (i.e `0.0.0.0/0` and `::/0`). This BANP will make it so user workloads can send traffic to other in-cluster endpoints, but not to external endpoints by default. `NetworkPolicy` can then be used by developers in order to allow their applications to send traffic to an explicit set of external services.
```

Author: Thank you this is a good rewording.

* Be careful that you scope your BANP so that it only denies traffic in user namespaces and not to system namespaces. This is because the system namespaces do not have `NetworkPolicy` to override your BANP.

[id="differences-with-netpol_{context}"]
= Differences to consider between NetworkPolicy and ANP

> * Use workload specific labels to secure your workloads and then use those labels as the namespace selector instead of using `{}` string.

Reviewer 4: I'm lost on the workload specific labels.  Am I doing that with NetworkPolicy or ANP?

> * Use workload specific labels to secure your workloads and then use those labels as the namespace selector instead of using `{}` string.

Reviewer 3: ```suggestion
* Use explicit labels to reference your workloads within ANP and BANP rather than using the empty (`{}`) catch all selector to avoid accidental traffic selection.
```

[IMPORTANT]
====

> An empty namespace selector applied to a infrastructure namespace has the potential to leave your entire cluster unresponsive and in a non-functional state because

Reviewer: sentence trails off...

Author: I actually don't think we need to explain this. It is more straightforward to focus on the consequence and not the cause which is implicit.

====

> * API semantics for ANP are very different from `NetworkPolicy`. There is no implicit deny for ANP when you create the policy. You have to explicitly define what you need. As a consequence, you cannot use ANP to isolate a pod from everything for ingress traffic. You You might consider using `Deny` action to `0.0.0.0/0` and it will act like a default deny to everything or you could `Deny` from all other pods for ingress traffic.

Reviewer 3: ```suggestion
* API semantics for ANP are very different from `NetworkPolicy`. There is no implicit deny with ANP when you create the policy. You have to explicitly define what you need.
```
This Is a separate bullet that may need to go back up to the ANP section

```
ANP assumes that pods should NOT be directly accessible from outside the cluster, as a consequence, you cannot use ANP to allow/deny external ingress traffic.
```

Reviewer: > This is a separate bullet that may need to go back up to the ANP section
> 
>     ANP assumes that pods should NOT be directly accessible from outside the cluster, as a consequence, you cannot use ANP to allow/deny external ingress traffic.

Given that ovn-k doesn't allow external ingress anyway, this is a no-op and we don't need to mention it, right?

Author: I retitled the section and think we should keep this where it is rather than move it into the ANP section.

// Module included in the following assemblies:
//
// * list of assemblies where this module is included:
// networking/network_security/AdminNetworkPolicy/nw-ovn-k-anp-recommended-practices.adoc

:_mod-docs-content-type: REFERENCE
[id="anp-best-practices_{context}"]
= Designing AdminNetworkPolicy

When building `AdminNetworkPolicy` (ANP), you should consider the following when creating your policies:

//This will need to be added in at 4.16.1

> //* You should create `AdminNetworkPolicy` at different priorities so that precedence is well defined. It is not guaranteed which policy will take precedence when overlapping ANP are created.

Reviewer 2: In 4.16.3 there can be ANP at the same priority but admin needs to ensure the rules are not lapping.

[IMPORTANT]
====
Applying ANP and BANP to system namespaces (`default`, `kube-system`, any namespace whose name starts with `openshift-`, etc) is not supported, and this can leave your cluster unresponsive and in a non-functional state.
====

* Since `0-100` is the supported priority range, it is always better when designing your ANP to use a middle range like `30-70` because this leaves some placeholder
for priorities before and after. Even in the middle range, you might want to leave gaps so that as your infrastructure requirements evolve over time, you are able to insert new ANPs when needed at the right priority level. If you pack your ANPs, then you might need to recreate all of them to accommodate any changes in the future.

* When using `0.0.0.0/0` or `::/0` to create a strong `Deny` policy, it is recommended that you check to make sure that you have higher priority `Allow` or `Pass` rules for essential traffic.

* Use `Allow` when you want to ensure that a connection is allowed no matter what. An `Allow` rule in an ANP means that the connection will always be allowed, and `NetworkPolicy` will be ignored.

> * Use `Pass` if you want to "not deny" the connection, and allow the users to decide for themselves whether to allow or deny the connection according to their own `NetworkPolicy` objects.`Pass` delegates policy decision to the `NetworkPolicy` layer.

Reviewer 2: @JoeAldinger You may want elaborate on users here - the project owners or developers. They get to define network policy while ANP is created by administrator of the cluster.

Reviewer 2: It is being mentioned later in BANP section. Should mention here as well for the rule with Pass action.

// Module included in the following assemblies:
//
// * list of assemblies where this module is included:
// networking/network_security/AdminNetworkPolicy/nw-ovn-k-anp-recommended-practices.adoc

:_mod-docs-content-type: REFERENCE
[id="anp-best-practices_{context}"]
= Designing AdminNetworkPolicy

> When building `AdminNetworkPolicy` (ANP) resources, you should consider the following when creating your policies:

Author: These aren't actions that `must` be taken so I switched to `might` to indicate possibility. I'm going to keep building to avoid repetitive language.

* You should create `AdminNetworkPolicy` at different priorities so that precedence is well defined. It is not guaranteed which policy will take precedence when overlapping ANP are created.

* Administrators should build ANP that apply to user namespaces not system namespaces.

[IMPORTANT]
====
Applying ANP and BANP to system namespaces (`default`, `kube-system`, any namespace whose name starts with `openshift-`, etc) is not supported, and this can leave your cluster unresponsive and in a non-functional state.
====

* Since `0-100` is the supported priority range, it is always better when designing your ANP to use a middle range like `30-70` because this leaves some placeholder

> for priorities before and after. Even in the middle range, you might want to leave gaps so that as your infrastructure requirements evolve over time, you are able to insert new ANPs when needed at the right priority level. If you pack your ANPs, then you might need to recreate all of them to accommodate any changes in the future.

Author: I _think_ we do since this feature is so new and we are speaking about best practices which is a genre of prescriptive writing. I take your point though about the minimalism and trimmed some words.

* When using `0.0.0.0/0` or `::/0` to create a strong `Deny` policy, it is recommended that you check to make sure that you have higher priority `Allow` or `Pass` rules for essential traffic.

> * Use `Allow` when you want to ensure that a connection is allowed no matter what. An `Allow` rule in an ANP means that the connection will always be allowed, and `NetworkPolicy` will be ignored.

Reviewer 5: ```suggestion
* Use `Allow` when you want to ensure that a connection is always allowed irrespective of the`NetworkPolicy`.
```

Reviewer 5: is this value in a particular field?

> * Use `Pass` if you want to "not deny" the connection, and allow developers to decide for themselves whether to allow or deny the connection according to their own `NetworkPolicy` objects.`Pass` delegates policy decision to the `NetworkPolicy` layer.

Reviewer 5: I might not be suggesting the best sentence here but it can be minimal similar to below.
```suggestion
* Use `Pass` to delegate the policy decision of allowing or denying the connection to the `NetworkPolicy` layer.
```

Reviewer 5: Do we also want this to be a section or a separate bullet about list of values allowed by the field?

Author: That exists in a separate section of the docs.

> * Selectors across multiple rules should not overlap. This ensures that that the same IPs do not appear in multiple policies, which can cause performance and scale limitations.

Reviewer 5: ```suggestion
* Ensure that the selectors across multiple rules do not overlap so that the same IPs do not appear in multiple policies, which can cause performance and scale limitations.
```

> * Avoid using `namedPorts` in conjunction with `PortNumber` and `PortRange` creates 6 ACLs and cause inefficiencies in your cluster.

Reviewer 5: I see some issue in this sentence formation.

[id="considerations-using-banp_{context}"]
== Considerations for using BaselineAdminNetworkPolicy

> * Only a single `BaselineAdminNetworkPolicy` (BANP) resource may be defined within a cluster, make sure you architect it thoughtfully. The following are supported uses for BANP that administrators might consider in designing their BANP:

Reviewer 5: ```suggestion
* You can define only a single `BaselineAdminNetworkPolicy` (BANP) resource within a cluster. The following are supported uses for BANP that administrators might consider in designing their BANP:
```

> ** You can set a default deny policy for cluster-local ingress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the ingress traffic that they want to allow, and if they do not add network policies for ingress it will be denied.

Reviewer 5: Second sentence says `make it`, not sure what it implies. May be the whole sentence needs a bit of rephrase to sound more meaningful in present tense.

> ** You can set a default deny policy for cluster-local egress in user namespaces. This BANP will make it so that users have to add `NetworkPolicy` to allow the egress traffic that they want to allow, and if they do not add network policies it will be denied.

Reviewer 5: same comment as in ingress

> ** You can set a default allow policy for egress to the in-cluster DNS service. Such a BANP will ensure that namespaced users do not have to set an allow egress `NetworkPolicy` to the in-cluster DNS service.

Reviewer 5: I am not sure if `default allow policy` is a valid term. may be allow needs to be in backticks or the sentence s needs to use it like a verb? Same applies to deny.

```suggestion
** You can set a default allow policy for egress to the in-cluster DNS service. Such a BANP ensures that the namespaced users do not have to set an allow egress `NetworkPolicy` to the in-cluster DNS service.
```

> ** You can set an egress policy that allows internal egress traffic to all pods but denies access to all external endpoints (i.e `0.0.0.0/0` and `::/0`). This BANP will make it so user workloads can send traffic to other in-cluster endpoints, but not to external endpoints by default. `NetworkPolicy` can then be used by developers in order to allow their applications to send traffic to an explicit set of external services.

Reviewer 5: there's a usage of `make it` here which probably needs rephrase.

`** You can set an egress policy that allows internal egress traffic to all pods but denies access to all external endpoints, which are `0.0.0.0/0` and `::/0`. .
`

> * Be careful that you scope your BANP so that it only denies traffic in user namespaces and not to system namespaces. This is because the system namespaces do not have `NetworkPolicy` to override your BANP.

Reviewer 5: To keep it parallel, may be use `in` or `to` for both user and system namespaces.

```suggestion
* Ensure that you scope your BANP so that it only denies traffic in user namespaces and not in(?) system namespaces. This is because the system namespaces do not have `NetworkPolicy` to override your BANP.
```

[id="differences-with-netpol_{context}"]

> = Differences to consider between AdminNetworkPolicy and NetworkPolicy

Reviewer 5: This is currently heading level 1 which needs correcyion
```suggestion
= Differences between AdminNetworkPolicy and NetworkPolicy
```

> * Use explicit labels to reference your workloads within ANP and BANP rather than using the empty (`{}`) catch all selector to avoid accidental traffic selection.

Reviewer 5: You may want to highlight the difference here.

[IMPORTANT]
====

> An empty namespace selector applied to a infrastructure namespace has the potential to leave your entire cluster unresponsive and in a non-functional state.

Reviewer 5: ```suggestion
An empty namespace selector applied to a infrastructure namespace can make your cluster unresponsive and in a non-functional state.
```

====

> * API semantics for ANP are very different from `NetworkPolicy` objects. There is no implicit deny with ANP when you create the policy. You have to explicitly define what you need.

Reviewer 5: again, the below sentence might need a revisit on the right usage of allow and deny. Any blanket term like authorization?
```suggestion
* In API semantics for ANP , you have to explicitly define allow or deny (?) when you create the policy, unlike BANP
```

> * Unlike `NetworkPolicy` objects, `AdminNetworkPolicy` objects ingress rules are limited to in-cluster pods and namespaces so you cannot, and do not need to, set rules for ingress from the host network.

Reviewer 2: @JoeAldinger Nothing stops for using for host networked pods, just the ANP rules are not enforced on them. They are only enforced pods created on cluster network.

Reviewer 2: Rest of it looks good.
I did not see reference to ANP at same priority with non lapping rules.

Reviewer 2: @palonsoro Are you good with wording of line number 61. If you are I will be OK to apply lgtm label.

Author: I worked with him over slack to develop that wording. I can message you the conversation if you'd like. He's out of office.

Reviewer 6: Yes. The idea is not to mean that we can't have host networked pod. The idea is to make it clear that host networked pods are not impacted (i.e. they are not limited) by ANPs and BANPs.

If you believe that the wording leads to confusion in this regard, not sure if you could suggest some alternative. But to me, we can leave it as is.

Author: I _think_ `do not need to` gets at what we are trying to express.

> * Unlike `NetworkPolicy` objects, `AdminNetworkPolicy` objects ingress rules are limited to in-cluster pods and namespaces so you cannot, and do not need to, set rules for ingress from the host network.

Reviewer 5: this seems to be difference between NP and ANP (?) which doesn't match the title. The title might have to be reconsidered if all the above points  needs to be incorporated.
